### PR TITLE
Add matrix decomposition functions, any, all, gaussian process example

### PR DESCRIPTION
**Pull request by austinvhuang**
- matrix decomposition / solvers
- any and all logical reductions
- porting gaussian process example (WIP)

### CODE REVIEW

**Comment by austinvhuang** [comment]
This should be ready for merging, feel free to review anyone.

File: examples/gaussian_process/Main.hs [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        t' = asTensor (snd <$> rngPairs)
+        pairs axis1' axis2' = [(t, t') | t <- axis1', t' <- axis2']
+        rngPairs = pairs axis1 axis2
+
+-- | 1-dimensional radial basis function kernel
+kernel1d_rbf :: Double -> Double -> Tensor -> Tensor -> Tensor
+kernel1d_rbf sigma length t t' = (sigma'^2) * exp eterm
+    where
+        sigma' = asTensor sigma
+        eterm = cmul (- (pow (t - t') (2 :: Int))) (1 / 2 * length^2)
```
Comment (junjihashimoto) [inline examples/gaussian_process/Main.hs]: Isn't it  `(1 / (2 * length^2) )` ?

**Review by junjihashimoto — COMMENTED** [review]
I want to check some details, but everything else is fine.

File: examples/gaussian_process/Main.hs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+-- | Compute posterior mean and covariance parameters based on observed data y
+condition :: Tensor -> Tensor -> Tensor -> Tensor -> Tensor -> Tensor -> (Tensor, Tensor)
+condition muX muY covXX covXY covYY y =
+    (postMu, postCov)
+    where
+        covYX = transpose2D covXY
+        invY = inverse covYY
+        postMu = muX + (matmul covXY (matmul invY (y - muY)))
+        postCov = covXX - (matmul covXY (matmul invY covYX))
```
Comment (junjihashimoto) [inline examples/gaussian_process/Main.hs:56]: The precision of `gesv` or `posv` may be better than  `inverse`'s one.
The code becomes a bit complicated.

Comment (austinvhuang) [reply examples/gaussian_process/Main.hs:56]: Do you know what those map to in the libtorch API? The raw lapack library isn't surfaced in the C++ API. Searching the pytorch repo, the closest I can find is https://github.com/pytorch/pytorch/blob/054c7eb0f4e854e764f2dd4aa727fbb575ee02b8/aten/src/ATen/native/BatchLinearAlgebra.cpp that doesn't seem to correspond to functions surfaced in the spec.

This not being a full-blown gp library yet, i can accept an okay-ish solution if it keeps the code and library usage simpler.

Comment (junjihashimoto) [reply examples/gaussian_process/Main.hs:56]: Sorry for not checking those map. (PR is acceptable.)
Those map is something complicated, and `posv` does not exist.

libtorch | lapack/blas
---------|---
gels     | gels
solve    | [sd]gesv
triangular_solve | [sd]trtrs
inverse | [sd]getri
cholesky_solve | [sd]potrs
cholesky | [sd]potrf
pstrf | [sd]pstrf
qr | qr
geqrf | geqrf
orgqr | orgqr
ormqr | ormqr

**Review by junjihashimoto — APPROVED** [review]

**Pull request merged**